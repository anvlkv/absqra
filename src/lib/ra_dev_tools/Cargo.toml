[package]
name = "ra_dev_tools"
version = "0.1.0"
authors = ["a.nvlkv <a.nvlkv@protonmail.ch>"]
edition = "2018"
build = "build.rs"

[build-dependencies]
build-deps = "0.1.4"
# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
insta = "0.16.0"
regex = "1.3.9"
example_files_macro = {path="./src/lib/example_files_macro"}
files_iterator = {path="./src/lib/files_iterator"}
