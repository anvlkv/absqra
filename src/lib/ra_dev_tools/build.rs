extern crate build_deps;

fn main() {
    // Enumerate files in sub-folder "examples/*", being relevant for the test-generation (as example)
    // If function returns with error, exit with error message.
    build_deps::rerun_if_changed_paths( "../../../examples/*/*" ).unwrap();

    // Adding the parent directory "examples" to the watch-list will capture new-files being added
    build_deps::rerun_if_changed_paths( "../../../examples/*" ).unwrap();
}