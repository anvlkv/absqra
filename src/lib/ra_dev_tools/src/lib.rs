pub extern crate insta;
pub use files_iterator::for_each_ra_example_file;

#[allow(unused_attributes)]
#[macro_use] pub use example_files_macro::make_example_tests;

#[cfg(test)]
mod test {
    use super::make_example_tests;

    #[make_example_tests]
    #[test]
    fn it_works_with(content: String, file_name: String) {
        println!("didn't panic with content: \n[{}] of file_name: {}", content, file_name);
    }
}