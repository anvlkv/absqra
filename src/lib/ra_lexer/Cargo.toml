[package]
name = "ra_lexer"
version = "0.1.0"
authors = ["a.nvlkv <a.nvlkv@protonmail.ch>"]
edition = "2018"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
failure = "0.1.8"
failure_derive = "0.1.8"
serde = "1.0.114"
serde_json = "1.0.56"
serde_derive = "1.0.114"

[dev-dependencies]
ra_dev_tools = {path = "../ra_dev_tools"}
