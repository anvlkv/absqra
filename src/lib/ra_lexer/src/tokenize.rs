use super::*;

pub fn tokenize<'a>(input: &'a str) -> impl Iterator<Item = Result<RaToken, LexerError>> + 'a {
    Cursor::from(input)
    // tokenize_cursor(Cursor::new(input, Position(1, 0), 0, 0)).map(|lexer_result| lexer_result?.try_into())
}