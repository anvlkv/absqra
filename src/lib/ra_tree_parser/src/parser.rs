use super::*;
use ra_lexer::errors::LexerError;

pub fn parse<'a>(
    mut tokens_stream: impl Iterator<Item = Result<RaToken, LexerError>>
) -> Result<RaTree, (Vec<LexerError>, Option<RaTree>)> {

    let mut tree = RaTree::new();

    let mut errors: Vec<LexerError> = Vec::new();

    while let Some(token_result) = tokens_stream.next() {
        match token_result {
            Ok(token) => {
                tree.add_token(token);
            },
            Err(e) => {
                errors.push(e);
            }
        }
    }
    if errors.len() > 0 {
        Err((errors, Some(tree)))
    }
    else {
        Ok(tree)
    }
}
