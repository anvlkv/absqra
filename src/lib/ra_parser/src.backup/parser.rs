use crate::expressions::context_expression::{ContextExpression, ContextExpressionMemberKind, ContextExpressionMember};
use ra_lexer::token::{RaToken, TokenKind};
use ra_lexer::tokenize;
use ra_lexer::errors::LexerError;

use super::block::{Block, BlockKind};
use super::cursor::Cursor;
use super::errors::ParserError;
use super::expressions::traits::{Leveled, Expandable, Positioned};
use failure::Backtrace;

pub fn parse<'a>(input: &'a str) -> Result<Block<'a>, (Vec<ParserError>, Block<'a>)> {
    let stream = tokenize(input);
    
    let mut errors: Vec<ParserError> = Vec::new();

    let program = {
        let mut block = Block::new(RaToken::default()).unwrap();

        let mut tokens_cursor = Cursor::new(
            stream
            .take_while(|r| r.is_ok())
            .map(|r| r.unwrap())
            .filter(|tok| tok.kind.unwrap() != TokenKind::Comment)
        );
        loop {
            if !tokens_cursor.is_eof() {
                match tokens_cursor.advance_block() {
                    Ok(blk) => block.children.push(blk),
                    Err(es) => errors.extend(es.into_iter()),
                }
            } else {
                break;
            }
        }

        block
    };

    if errors.len() > 0 {
        Err((errors, program))
    } else {
        Ok(program)
    }
}

impl<'token, I> Cursor<'token, I>
where
    I: Iterator<Item = RaToken<'token>>,
{
    fn advance_block<'a>(&'a mut self) -> Result<Block<'token>, Vec<ParserError>> {
        let first_token = self.bump();
        let mut errors: Vec<ParserError> = Vec::new();
        match first_token {
            Some(token) => {
                self.parse_block(token)
            },
            None => {
                errors.push(ParserError::UnexpectedEndOfInput(self.position, Backtrace::new()));
                Err(errors)
            }
        }
    }

    fn parse_block(
        &mut self,
        first_token: RaToken<'token>,
    ) -> Result<Block<'token>, Vec<ParserError>> {
        let mut errors: Vec<ParserError> = Vec::new();
        let block = Block::new(first_token)?;

        if let (Some(blk), mut errs) =
            self.check_parse_block_expression(block, &mut errors)
        {
            if let (Some(blk), mut errs) = self.check_parse_block_children(blk, &mut errs) {
                if let (Some(blk), errs) = self.check_parse_union(blk, &mut errs) {
                    if errs.len() == 0 {
                        return Ok(blk);
                    }
                }
            }
        }

        Err(errors)
    }

    fn check_parse_block_expression<'errors>(
        &mut self,
        mut block: Block<'token>,
        errors: &'errors mut Vec<ParserError>,
    ) -> (Option<Block<'token>>, &'errors mut Vec<ParserError>) {
        let mut lvl = block.get_level();
        let mut line_number = (block.get_position().0).0;
        while self.first_ahead().is_some()
            && self.first_ahead().unwrap().level == lvl
            && (self.first_ahead().unwrap().position.0).0 == line_number {
                match block.clone().append_item(self.bump().unwrap()) {
                    Ok(blk) => {
                        block = blk;

                        // // TODO: ugly, move this logic elsewhere

                        if let BlockKind::ContextModification(Some(ContextExpression(target, source))) = block.clone().kind { 
                            if ContextExpressionMember::Target(ContextExpressionMemberKind::MSpecifier) == target && source.is_none() {
                                if let (Some(mut blk), mut errs) = self.check_parse_block_children(block.clone(), errors) {
                                    let close_curly_brace = self.bump().unwrap();
                                    assert_eq!(close_curly_brace.kind.unwrap(), TokenKind::CloseCurlyBrace);
                                    block = blk;
                                    lvl = close_curly_brace.level;
                                    line_number = (close_curly_brace.position.0).0;
                            //         // let updated_expression = ContextExpression(target, None).append_item(self.bump().unwrap())?;

                            //         // blk.kind = BlockKind::ContextModification(Some(updated_expression));
                            //         // self.check_parse_block_expression(blk, errs)
                                }
                            }
                            else if source.is_some() && ContextExpressionMember::Source(ContextExpressionMemberKind::MSpecifier) == source.unwrap() {
                                if let (Some(blk), mut errs) = self.check_parse_block_children(block.clone(), errors) {
                                    assert_eq!(self.bump().unwrap().kind.unwrap(), TokenKind::CloseCurlyBrace);
                                    assert_eq!(self.bump().unwrap().kind.unwrap(), TokenKind::CloseCurlyBrace);
                                    block = blk;
                                    // return Ok(blk);
                                }
                            }
                            println!("{:?}", block);
                        }

                        // // end TODO
                    },
                    Err(e) => {
                        errors.push(e)
                    }
                }
        }

        (Some(block), errors)
    }

    fn check_parse_block_children<'errors>(
        &mut self,
        mut block: Block<'token>,
        errors: &'errors mut Vec<ParserError>,
    ) -> (Option<Block<'token>>, &'errors mut Vec<ParserError>) {
        
        while !self.is_eof() && self.first_ahead().unwrap().level == block.get_level() + 1 {
            match self.advance_block() {
                Ok(child) => {
                    block.children.push(child);
                }
                Err(errs) => {
                    errors.extend(errs);
                }
            }
        }

        if errors.len() == 0 {
            (Some(block), errors)
        } else {
            (None, errors)
        }
    }

    fn check_parse_union<'errors>(
        &mut self,
        block: Block<'token>,
        errors: &'errors mut Vec<ParserError>,
    ) -> (Option<Block<'token>>, &'errors mut Vec<ParserError>) {
        let mut union_block = block.clone();
        let lvl = block.get_level();
        let mut union_size = 0;
        while !self.is_eof()
            && self.first_ahead().unwrap().kind.unwrap() == TokenKind::Coma
            && self.first_ahead().unwrap().level == lvl
        {
            assert!(self.bump().unwrap().kind.unwrap() == TokenKind::Coma);
            match self.advance_block() {
                Ok(blk) => {
                    union_size += 1;
                    union_block.children.push(blk);
                }
                Err(errs) => {
                    errors.extend(errs);
                }
            }
        }
        
        if union_size > 0 {
            union_block.kind = BlockKind::Union(union_size);
            (Some(union_block), errors)
        } else {
            (Some(block), errors)
        }
    }
}
