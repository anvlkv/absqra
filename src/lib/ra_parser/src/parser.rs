use crate::errors::ParserError;
use crate::block::Block;
use crate::parsed_by_token::ParsedByToken;
use ra_lexer::tokenize;
use ra_lexer::token::TokenKind;

pub fn parse<'a>(input: &'a str) -> Result<Block, (Vec<ParserError>, Option<Block>)> {
    let mut errors: Vec<ParserError> = Vec::new();
    let mut tokens_stream = tokenize(input);
    
    let mut block = {
        match Block::new_program() {
            Ok(b) => b,
            Err(e) => {
                errors.extend(e);
                return Err((errors, None));
            }
        }
    };

    while let Some(token_result) = tokens_stream.next() {
        match token_result {
            Ok(token) => {
                if token.kind == TokenKind::Comment {
                    continue;
                }
                else {
                    match block.clone().append_token(token) {
                        Ok(blk) => block=blk,
                        Err(e) => errors.extend(e)
                    }
                }
            },
            Err(e) => {
                errors.push(e.into());
            }
        }
    }

    if errors.len() == 0 {
        Ok(*block)
    }
    else {
        Err((errors, Some(*block)))
    }
}