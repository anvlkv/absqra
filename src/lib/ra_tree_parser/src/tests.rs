
mod lib {
    use ra_dev_tools::insta::{assert_snapshot, assert_debug_snapshot};    
    use ra_dev_tools::{make_example_tests};
    use ra_lexer::tokenize;
    use crate::parser::parse;

    #[make_example_tests]
    #[test]
    fn it_should_match_snapshots(contents: String, file_name: String) {
        let err_messages_file_name = format!("{}__{}",file_name, "ERR__MESSAGES");
        let errors_file_name = format!("{}__{}", file_name, "ERR");
        let tree = {
            let mut tokens_iter = tokenize(&contents);
            match parse(tokens_iter) {
                Ok(tree) => {
                    tree
                },
                Err((errors, parsed)) => {
                    let formatted_errors: String = {
                        errors.into_iter().map(|e| format!("{}", e)).collect::<Vec<String>>().join("\n")
                    };
                    
                    assert_snapshot!(err_messages_file_name, formatted_errors);
                    assert_debug_snapshot!(errors_file_name, parsed);


                    panic!("failed to parse example {:?}", file_name);
                }
            }
        };
        assert_debug_snapshot!(file_name, tree)       
    }
}
