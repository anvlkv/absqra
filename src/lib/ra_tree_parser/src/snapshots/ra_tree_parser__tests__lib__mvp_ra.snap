---
source: src/lib/ra_tree_parser/src/tests.rs
expression: tree
---
{
  "tokens": [],
  "level": 0,
  "position": [
    [
      0,
      0
    ],
    [
      32,
      26
    ]
  ],
  "children": [
    {
      "tokens": [
        {
          "kind": {
            "Comment": [
              " this file is my todo list and also should render like one",
              false
            ]
          },
          "position": [
            [
              1,
              1
            ],
            [
              1,
              61
            ]
          ],
          "level": 1
        }
      ],
      "level": 1,
      "position": [
        [
          1,
          1
        ],
        [
          1,
          61
        ]
      ],
      "children": []
    },
    {
      "tokens": [
        {
          "kind": {
            "Identifier": "parseBlocks"
          },
          "position": [
            [
              3,
              1
            ],
            [
              3,
              12
            ]
          ],
          "level": 1
        }
      ],
      "level": 1,
      "position": [
        [
          3,
          1
        ],
        [
          10,
          16
        ]
      ],
      "children": [
        {
          "tokens": [
            {
              "kind": "Greater",
              "position": [
                [
                  4,
                  5
                ],
                [
                  4,
                  6
                ]
              ],
              "level": 2
            }
          ],
          "level": 2,
          "position": [
            [
              4,
              5
            ],
            [
              10,
              16
            ]
          ],
          "children": [
            {
              "tokens": [
                {
                  "kind": "Exclamation",
                  "position": [
                    [
                      5,
                      9
                    ],
                    [
                      5,
                      10
                    ]
                  ],
                  "level": 3
                },
                {
                  "kind": {
                    "Identifier": "choice"
                  },
                  "position": [
                    [
                      5,
                      10
                    ],
                    [
                      5,
                      16
                    ]
                  ],
                  "level": 3
                },
                {
                  "kind": "Colon",
                  "position": [
                    [
                      5,
                      16
                    ],
                    [
                      5,
                      17
                    ]
                  ],
                  "level": 3
                },
                {
                  "kind": {
                    "Identifier": "multiple"
                  },
                  "position": [
                    [
                      5,
                      17
                    ],
                    [
                      5,
                      25
                    ]
                  ],
                  "level": 3
                }
              ],
              "level": 3,
              "position": [
                [
                  5,
                  9
                ],
                [
                  5,
                  25
                ]
              ],
              "children": []
            },
            {
              "tokens": [
                {
                  "kind": "Exclamation",
                  "position": [
                    [
                      6,
                      9
                    ],
                    [
                      6,
                      10
                    ]
                  ],
                  "level": 3
                },
                {
                  "kind": {
                    "Identifier": "options"
                  },
                  "position": [
                    [
                      6,
                      10
                    ],
                    [
                      6,
                      17
                    ]
                  ],
                  "level": 3
                },
                {
                  "kind": "Colon",
                  "position": [
                    [
                      6,
                      17
                    ],
                    [
                      6,
                      18
                    ]
                  ],
                  "level": 3
                }
              ],
              "level": 3,
              "position": [
                [
                  6,
                  9
                ],
                [
                  10,
                  16
                ]
              ],
              "children": [
                {
                  "tokens": [
                    {
                      "kind": {
                        "ContentBlock": "parse blocks"
                      },
                      "position": [
                        [
                          7,
                          13
                        ],
                        [
                          7,
                          27
                        ]
                      ],
                      "level": 4
                    }
                  ],
                  "level": 4,
                  "position": [
                    [
                      7,
                      13
                    ],
                    [
                      7,
                      27
                    ]
                  ],
                  "children": []
                },
                {
                  "tokens": [
                    {
                      "kind": {
                        "ContentBlock": "and funny chars too ðŸ˜ "
                      },
                      "position": [
                        [
                          8,
                          13
                        ],
                        [
                          8,
                          36
                        ]
                      ],
                      "level": 4
                    }
                  ],
                  "level": 4,
                  "position": [
                    [
                      8,
                      13
                    ],
                    [
                      8,
                      36
                    ]
                  ],
                  "children": []
                },
                {
                  "tokens": [
                    {
                      "kind": {
                        "ContentBlock": "â‚°"
                      },
                      "position": [
                        [
                          9,
                          13
                        ],
                        [
                          9,
                          16
                        ]
                      ],
                      "level": 4
                    }
                  ],
                  "level": 4,
                  "position": [
                    [
                      9,
                      13
                    ],
                    [
                      9,
                      16
                    ]
                  ],
                  "children": []
                },
                {
                  "tokens": [
                    {
                      "kind": {
                        "ContentBlock": "Ï¿"
                      },
                      "position": [
                        [
                          10,
                          13
                        ],
                        [
                          10,
                          16
                        ]
                      ],
                      "level": 4
                    }
                  ],
                  "level": 4,
                  "position": [
                    [
                      10,
                      13
                    ],
                    [
                      10,
                      16
                    ]
                  ],
                  "children": []
                }
              ]
            }
          ]
        }
      ]
    },
    {
      "tokens": [
        {
          "kind": {
            "Identifier": "parseDifferentTypesOfBlocks"
          },
          "position": [
            [
              12,
              1
            ],
            [
              12,
              28
            ]
          ],
          "level": 1
        }
      ],
      "level": 1,
      "position": [
        [
          12,
          1
        ],
        [
          22,
          28
        ]
      ],
      "children": [
        {
          "tokens": [
            {
              "kind": "Greater",
              "position": [
                [
                  13,
                  5
                ],
                [
                  13,
                  6
                ]
              ],
              "level": 2
            }
          ],
          "level": 2,
          "position": [
            [
              13,
              5
            ],
            [
              22,
              28
            ]
          ],
          "children": [
            {
              "tokens": [
                {
                  "kind": "Exclamation",
                  "position": [
                    [
                      14,
                      9
                    ],
                    [
                      14,
                      10
                    ]
                  ],
                  "level": 3
                },
                {
                  "kind": {
                    "Identifier": "choice"
                  },
                  "position": [
                    [
                      14,
                      10
                    ],
                    [
                      14,
                      16
                    ]
                  ],
                  "level": 3
                },
                {
                  "kind": "Colon",
                  "position": [
                    [
                      14,
                      16
                    ],
                    [
                      14,
                      17
                    ]
                  ],
                  "level": 3
                },
                {
                  "kind": {
                    "Identifier": "multiple"
                  },
                  "position": [
                    [
                      14,
                      17
                    ],
                    [
                      14,
                      25
                    ]
                  ],
                  "level": 3
                }
              ],
              "level": 3,
              "position": [
                [
                  14,
                  9
                ],
                [
                  14,
                  25
                ]
              ],
              "children": []
            },
            {
              "tokens": [
                {
                  "kind": "Exclamation",
                  "position": [
                    [
                      15,
                      9
                    ],
                    [
                      15,
                      10
                    ]
                  ],
                  "level": 3
                },
                {
                  "kind": {
                    "Identifier": "options"
                  },
                  "position": [
                    [
                      15,
                      10
                    ],
                    [
                      15,
                      17
                    ]
                  ],
                  "level": 3
                },
                {
                  "kind": "Colon",
                  "position": [
                    [
                      15,
                      17
                    ],
                    [
                      15,
                      18
                    ]
                  ],
                  "level": 3
                }
              ],
              "level": 3,
              "position": [
                [
                  15,
                  9
                ],
                [
                  22,
                  28
                ]
              ],
              "children": [
                {
                  "tokens": [
                    {
                      "kind": {
                        "ContentBlock": "comments"
                      },
                      "position": [
                        [
                          16,
                          13
                        ],
                        [
                          16,
                          23
                        ]
                      ],
                      "level": 4
                    }
                  ],
                  "level": 4,
                  "position": [
                    [
                      16,
                      13
                    ],
                    [
                      16,
                      23
                    ]
                  ],
                  "children": []
                },
                {
                  "tokens": [
                    {
                      "kind": {
                        "ContentBlock": "var declaration"
                      },
                      "position": [
                        [
                          17,
                          13
                        ],
                        [
                          17,
                          30
                        ]
                      ],
                      "level": 4
                    }
                  ],
                  "level": 4,
                  "position": [
                    [
                      17,
                      13
                    ],
                    [
                      17,
                      30
                    ]
                  ],
                  "children": []
                },
                {
                  "tokens": [
                    {
                      "kind": {
                        "ContentBlock": "rule declaration"
                      },
                      "position": [
                        [
                          18,
                          13
                        ],
                        [
                          18,
                          31
                        ]
                      ],
                      "level": 4
                    }
                  ],
                  "level": 4,
                  "position": [
                    [
                      18,
                      13
                    ],
                    [
                      18,
                      31
                    ]
                  ],
                  "children": []
                },
                {
                  "tokens": [
                    {
                      "kind": {
                        "ContentBlock": "applied rule"
                      },
                      "position": [
                        [
                          19,
                          13
                        ],
                        [
                          19,
                          27
                        ]
                      ],
                      "level": 4
                    }
                  ],
                  "level": 4,
                  "position": [
                    [
                      19,
                      13
                    ],
                    [
                      19,
                      27
                    ]
                  ],
                  "children": []
                },
                {
                  "tokens": [
                    {
                      "kind": {
                        "ContentBlock": "input block"
                      },
                      "position": [
                        [
                          20,
                          13
                        ],
                        [
                          20,
                          26
                        ]
                      ],
                      "level": 4
                    }
                  ],
                  "level": 4,
                  "position": [
                    [
                      20,
                      13
                    ],
                    [
                      20,
                      26
                    ]
                  ],
                  "children": []
                },
                {
                  "tokens": [
                    {
                      "kind": {
                        "ContentBlock": "context switches"
                      },
                      "position": [
                        [
                          21,
                          13
                        ],
                        [
                          21,
                          31
                        ]
                      ],
                      "level": 4
                    }
                  ],
                  "level": 4,
                  "position": [
                    [
                      21,
                      13
                    ],
                    [
                      21,
                      31
                    ]
                  ],
                  "children": []
                },
                {
                  "tokens": [
                    {
                      "kind": {
                        "ContentBlock": "content block"
                      },
                      "position": [
                        [
                          22,
                          13
                        ],
                        [
                          22,
                          28
                        ]
                      ],
                      "level": 4
                    }
                  ],
                  "level": 4,
                  "position": [
                    [
                      22,
                      13
                    ],
                    [
                      22,
                      28
                    ]
                  ],
                  "children": []
                }
              ]
            }
          ]
        }
      ]
    },
    {
      "tokens": [
        {
          "kind": {
            "Identifier": "generateRuntime"
          },
          "position": [
            [
              24,
              1
            ],
            [
              24,
              16
            ]
          ],
          "level": 1
        }
      ],
      "level": 1,
      "position": [
        [
          24,
          1
        ],
        [
          32,
          26
        ]
      ],
      "children": [
        {
          "tokens": [
            {
              "kind": "Greater",
              "position": [
                [
                  25,
                  5
                ],
                [
                  25,
                  6
                ]
              ],
              "level": 2
            }
          ],
          "level": 2,
          "position": [
            [
              25,
              5
            ],
            [
              32,
              26
            ]
          ],
          "children": [
            {
              "tokens": [
                {
                  "kind": "Exclamation",
                  "position": [
                    [
                      26,
                      9
                    ],
                    [
                      26,
                      10
                    ]
                  ],
                  "level": 3
                },
                {
                  "kind": {
                    "Identifier": "choice"
                  },
                  "position": [
                    [
                      26,
                      10
                    ],
                    [
                      26,
                      16
                    ]
                  ],
                  "level": 3
                },
                {
                  "kind": "Colon",
                  "position": [
                    [
                      26,
                      16
                    ],
                    [
                      26,
                      17
                    ]
                  ],
                  "level": 3
                },
                {
                  "kind": {
                    "Identifier": "multiple"
                  },
                  "position": [
                    [
                      26,
                      17
                    ],
                    [
                      26,
                      25
                    ]
                  ],
                  "level": 3
                }
              ],
              "level": 3,
              "position": [
                [
                  26,
                  9
                ],
                [
                  26,
                  25
                ]
              ],
              "children": []
            },
            {
              "tokens": [
                {
                  "kind": "Exclamation",
                  "position": [
                    [
                      27,
                      9
                    ],
                    [
                      27,
                      10
                    ]
                  ],
                  "level": 3
                },
                {
                  "kind": {
                    "Identifier": "options"
                  },
                  "position": [
                    [
                      27,
                      10
                    ],
                    [
                      27,
                      17
                    ]
                  ],
                  "level": 3
                },
                {
                  "kind": "Colon",
                  "position": [
                    [
                      27,
                      17
                    ],
                    [
                      27,
                      18
                    ]
                  ],
                  "level": 3
                }
              ],
              "level": 3,
              "position": [
                [
                  27,
                  9
                ],
                [
                  32,
                  26
                ]
              ],
              "children": [
                {
                  "tokens": [
                    {
                      "kind": {
                        "ContentBlock": "render content"
                      },
                      "position": [
                        [
                          28,
                          13
                        ],
                        [
                          28,
                          29
                        ]
                      ],
                      "level": 4
                    }
                  ],
                  "level": 4,
                  "position": [
                    [
                      28,
                      13
                    ],
                    [
                      28,
                      29
                    ]
                  ],
                  "children": []
                },
                {
                  "tokens": [
                    {
                      "kind": {
                        "ContentBlock": "render inputs"
                      },
                      "position": [
                        [
                          29,
                          13
                        ],
                        [
                          29,
                          28
                        ]
                      ],
                      "level": 4
                    }
                  ],
                  "level": 4,
                  "position": [
                    [
                      29,
                      13
                    ],
                    [
                      29,
                      28
                    ]
                  ],
                  "children": []
                },
                {
                  "tokens": [
                    {
                      "kind": {
                        "ContentBlock": "group inputs as long as there no dependency on previous step"
                      },
                      "position": [
                        [
                          30,
                          13
                        ],
                        [
                          30,
                          75
                        ]
                      ],
                      "level": 4
                    }
                  ],
                  "level": 4,
                  "position": [
                    [
                      30,
                      13
                    ],
                    [
                      30,
                      75
                    ]
                  ],
                  "children": []
                },
                {
                  "tokens": [
                    {
                      "kind": {
                        "ContentBlock": "execute logic"
                      },
                      "position": [
                        [
                          31,
                          13
                        ],
                        [
                          31,
                          28
                        ]
                      ],
                      "level": 4
                    }
                  ],
                  "level": 4,
                  "position": [
                    [
                      31,
                      13
                    ],
                    [
                      31,
                      28
                    ]
                  ],
                  "children": []
                },
                {
                  "tokens": [
                    {
                      "kind": {
                        "ContentBlock": "apply rules"
                      },
                      "position": [
                        [
                          32,
                          13
                        ],
                        [
                          32,
                          26
                        ]
                      ],
                      "level": 4
                    }
                  ],
                  "level": 4,
                  "position": [
                    [
                      32,
                      13
                    ],
                    [
                      32,
                      26
                    ]
                  ],
                  "children": []
                }
              ]
            }
          ]
        }
      ]
    }
  ]
}
